## Downloading a Structure File from the PDB database

In [1]:
from Bio.PDB import *
import nglview as nv

pdbl = PDBList()
pdbl.retrieve_pdb_file('1FAT')

Structure exists: '/home/emmy/Desktop/310.ai/fa/1fat.cif' 


'/home/emmy/Desktop/310.ai/fa/1fat.cif'

## Visualizing with the CIF Format

In [3]:
cif_parser = MMCIFParser()
structure = cif_parser.get_structure("1FAT", "fa/1fat.cif")

/home/emmy/anaconda3/lib/python3.9/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 7168.
  warnings.warn(
/home/emmy/anaconda3/lib/python3.9/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 7184.
  warnings.warn(
/home/emmy/anaconda3/lib/python3.9/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 7200.
  warnings.warn(
/home/emmy/anaconda3/lib/python3.9/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain D is discontinuous at line 7216.
  warnings.warn(
/home/emmy/anaconda3/lib/python3.9/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 7232.
  warnings.warn(
/home/emmy/anaconda3/lib/python3.9/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 723

In [4]:
def cleandir(obj):
    print(", ".join([a for a in dir(obj) if not a.startswith("_")]))
cleandir(structure)

add, atom_to_internal_coordinates, center_of_mass, child_dict, child_list, copy, detach_child, detach_parent, full_id, get_atoms, get_chains, get_full_id, get_id, get_iterator, get_level, get_list, get_models, get_parent, get_residues, has_id, header, id, insert, internal_to_atom_coordinates, level, parent, set_parent, transform, xtra


In [8]:
view = nv.show_biopython(structure)
view

NGLWidget()

## Accessing Protein Information

In [17]:
# Header
mmcif_dict = MMCIF2Dict.MMCIF2Dict("fa/1fat.cif")
len(mmcif_dict)

689

This generates a large dictionary of information about the protein, including the citation that sequenced the protein, structure information, atom by atom location and angles, and chemical compositions. As you can see, there are 689 items in this dictionary.

The Structure object in our file follows the SMCRA architecture in a parent-child relationship:

- A Structure consists of models
- A Model consists of chains
- A Chain consists of residues (amino acids)
- A Residue consists of Atoms

In [24]:
# Building Polypeptides

polypeptide_builder = CaPPBuilder()
counter = 1

for polypeptide in polypeptide_builder.build_peptides(structure):
    seq = polypeptide.get_sequence()
    print(f"Sequence: {counter}, Length: {len(seq)}")
    print(seq, "\n")
    counter += 1

Sequence: 1, Length: 36
SNDIYFNFQRFNETNLILQRDASVSSSGQLRLTNLN 

Sequence: 2, Length: 196
NGEPRVGSLGRAFYSAPIQIWDNTTGTVASFATSFTFNIQVPNNAGPADGLAFALVPVGSQPKDKGGFLGLFDGSNSNFHTVAVEFDTLYNKDWDPTERHIGIDVNSIRSIKTTRWDFVNGENAEVLITYDSSTNLLVASLVYPSQKTSFIVSDTVDLKSVLPEWVSVGFSATTGINKGNVETNDVLSWSFASKLS 

Sequence: 3, Length: 233
SNDIYFNFQRFNETNLILQRDASVSSSGQLRLTNLNGNGEPRVGSLGRAFYSAPIQIWDNTTGTVASFATSFTFNIQVPNNAGPADGLAFALVPVGSQPKDKGGFLGLFDGSNSNFHTVAVEFDTLYNKDWDPTERHIGIDVNSIRSIKTTRWDFVNGENAEVLITYDSSTNLLVASLVYPSQKTSFIVSDTVDLKSVLPEWVSVGFSATTGINKGNVETNDVLSWSFASKLS 

Sequence: 4, Length: 36
SNDIYFNFQRFNETNLILQRDASVSSSGQLRLTNLN 

Sequence: 5, Length: 196
NGEPRVGSLGRAFYSAPIQIWDNTTGTVASFATSFTFNIQVPNNAGPADGLAFALVPVGSQPKDKGGFLGLFDGSNSNFHTVAVEFDTLYNKDWDPTERHIGIDVNSIRSIKTTRWDFVNGENAEVLITYDSSTNLLVASLVYPSQKTSFIVSDTVDLKSVLPEWVSVGFSATTGINKGNVETNDVLSWSFASKLS 

Sequence: 6, Length: 35
SNDIYFNFQRFNETNLILQRDASVSSSGQLRLTNL 

Sequence: 7, Length: 196
NGEPRVGSLGRAFYSAPIQIWDNTTGTVASFATSFTFNIQVPNNAGPADGLAFALVPVGSQPKDKGGFLGLFDGSNSN

# Analyzing Residues Sequences

In [26]:
from Bio.SeqUtils.ProtParam import ProteinAnalysis

In [27]:
analyzed_seq = ProteinAnalysis(str(seq))